### PR TITLE
Fix Datadog pipeline options scope

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

Fixes the scope of `datadog` pipeline options:
- test visibility auto-instrumentation is only applied to the stages where `datadog` option is present
- custom tags are only applied to the stages where `datadog` option is present

See examples for better understanding.

In the below pipeline test visibility auto-instrumentation and "myTag" custom tag will be applied for every stage:
```
pipeline {
    agent any
    options {
        datadog(
            testVisibility: [ enabled: true, ...],
            tags: ["myTag:myValue"]
        )
    }
    stages {
        stage('Stage 1') {
            steps {
                ...
            }
        }
        
        stage('Stage 2') {
            steps {
                ...
            }
        }
        
        stage('Stage 3') {
            steps {
                ...
            }
        }
    }
}
```

In the below pipeline test visibility auto-instrumentation and "myTag" custom tag will be applied _only for Stage 2_:
```
pipeline {
    agent any
    stages {
        stage('Stage 1') {
            steps {
                ...
            }
        }
        
        stage('Stage 2') {
            options {
                datadog(
                    testVisibility: [ enabled: true, ...],
                    tags: ["myTag:myValue"]
                )
            }
            steps {
                ...
            }
        }
        
        stage('Stage 3') {
            steps {
                ...
            }
        }
    }
}
```

In the below pipeline test visibility auto-instrumentation and "myTag" custom tag will be applied for Stage 2, Stage 2.1, Stage 2.2, Stage 2.3.
Additionally, "myOtherTag" custom tag will be applied to Stage 2.2.
```
pipeline {
    agent any
    stages {
        stage('Stage 1') {
            steps {
                ...
            }
        }
        
        stage('Stage 2') {
            options {
                datadog(
                    testVisibility: [ enabled: true, ...],
                    tags: ["myTag:myValue"]
                )
            }
            stages {
                stage('Stage 2.1') {
                    steps {
                        ...
                    }
                }

                stage('Stage 2.2') {
                    options {
                        datadog(tags: ["myOtherTag:myOtherValue"])
                    }
                    steps {
                        ...
                    }
                }

                stage('Stage 2.3') {
                    steps {
                        ...
                    }
                }
            }
        }
        
        stage('Stage 3') {
            steps {
                ...
            }
        }
    }
}
```

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/steps/TestVisibility.java [hunk 1]
```diff
@@ -18,6 +18,13 @@ public class TestVisibility implements Serializable {
     public TestVisibility() {
     }
 
+    public TestVisibility(boolean enabled, String serviceName, Collection<TracerLanguage> languages, Map<String, String> additionalVariables) {
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/steps/TestVisibility.java:21]: Should we name this `TestOptimization` or `TestOptimizationConfig` class? (Given the fact that's just a model class with getters/setters that keeps the config of Test Optimization)

File: src/test/resources/org/datadog/jenkins/plugins/datadog/apm/test-multistage-maven-pipeline-with-datadog-step.txt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            steps {
+                sh '''
+                env | grep javaagent && echo "${STAGE_NAME} has tracer" || echo "${STAGE_NAME} has no tracer"
+                '''
+            }
+        }
+
+        stage('Stage 2') {
+            options {
+              datadog(testVisibility: [ enabled: true, serviceName: "my-service", languages: ["JAVA"], additionalVariables: ["my-var": "value"] ])
```
Comment (drodriguezhdez) [inline src/test/resources/org/datadog/jenkins/plugins/datadog/apm/test-multistage-maven-pipeline-with-datadog-step.txt:16]:
```suggestion
              datadog(testOptimization: [ enabled: true, serviceName: "my-service", languages: ["JAVA"], additionalVariables: ["my-var": "value"] ])
```

Should we rename this?

File: src/test/resources/org/datadog/jenkins/plugins/datadog/listeners/testPipelineStageTags.txt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        stage('Stage 1') {
+            steps {
+                echo 'Stage 1'
+            }
+        }
+
+        stage('Stage 2') {
+            options {
+                datadog(tags: ["outer_tag:value"])
+            }
```
Comment (drodriguezhdez) [inline src/test/resources/org/datadog/jenkins/plugins/datadog/listeners/testPipelineStageTags.txt:15]: Is it possible to have (and if so, what happens if):
- `tags` at root level and `tags` at the stage level at the same time?
- `tags` at root level with the same key than the `tags` at the stage level at the same time?

**Review by drodriguezhdez — COMMENTED** [review]
Dropped some questions about some scenarios that combines `options` both at root and at stage level. Finally, also dropped some comments about the complete renaming from `TestVisibility` to `TestOptimization`.

File: src/test/resources/org/datadog/jenkins/plugins/datadog/apm/test-multistage-maven-pipeline-with-datadog-step.txt (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply src/test/resources/org/datadog/jenkins/plugins/datadog/apm/test-multistage-maven-pipeline-with-datadog-step.txt]: Renamed

File: src/test/resources/org/datadog/jenkins/plugins/datadog/listeners/testPipelineStageTags.txt (hunk 3 above)
Comment (nikita-tkachenko-datadog) [reply src/test/resources/org/datadog/jenkins/plugins/datadog/listeners/testPipelineStageTags.txt]: > tags at root level and tags at the stage level at the same time?
yes, this is possible

> tags at root level with the same key than the tags at the stage level at the same time?
if the same tag is defined at multiple levels (root, stage, "substage"), the innermost value will be used. I actually had to fix this

**Review by drodriguezhdez — DISMISSED** [review]

**Review by drodriguezhdez — APPROVED** [review]

**Pull request merged**